File: py/graph/maximum_flow/SAP.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys
from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class SAP:
    def search(self, graphs: List[VertexNode], start: int, end: int) -> int:
        weights = [[0 for _ in range(len(graphs))] for _ in range(len(graphs))]
        for i, value in enumerate(graphs):
            edge = value.next
            while edge:
                weights[i][edge.index] = edge.weight
                edge = edge.next

        gap, degrees, result = [0 for _ in range(len(graphs) + 1)], [0 for _ in range(len(graphs))], 0
        gap[0] = len(graphs)
        while degrees[start] != len(graphs):
            result += self.__search(graphs, start, end, start, sys.maxsize, weights, gap, degrees)
        return result

    def __search(self, graphs: List[VertexNode], start: int, end: int, index: int, flow: int, weights: List[List[int]], gap: List[int], degrees: List[int]) -> int:
        if index == end:
            return flow

        edge, pivot, border = graphs[index].next, flow, len(graphs) - 1
        while edge:
            edgeIndex = edge.index
            if weights[index][edgeIndex]:
                if degrees[index] == degrees[edgeIndex] + 1:
                    value = self.__search(graphs, start, end, edgeIndex, min(pivot, weights[index][edgeIndex]), weights, gap, degrees)
                    weights[index][edgeIndex] -= value
                    weights[edgeIndex][index] += value
                    pivot -= value
                    if degrees[start] == len(graphs) or not pivot:
                        return flow - pivot
                if border > degrees[edgeIndex]:
                    border = degrees[edgeIndex]
            edge = edge.next

        gap[degrees[index]] -= 1
        if not ~gap[degrees[index]]:
            degrees[start] = len(graphs)
        else:
            degrees[index] = border + 1
            gap[degrees[index]] += 1
        return flow - pivot

File: cpp/graph/maximum_flow/SAP.cpp
#include <iostream>
#include <vector>
using namespace std;

struct EdgeNode {
    int index;
    int weight;
    EdgeNode *next;

    EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
};

struct VertexNode {
    int value;
    EdgeNode *next;

    VertexNode(int value) : value(value), next(nullptr) {}
};

class SAP {
public:
    int search(vector<VertexNode *> &graphs, int start, int end) {
        vector<vector<int>> weights(graphs.size(), vector<int>(graphs.size(), 0));
        for (int i = 0; i < (int) graphs.size(); ++i) {
            EdgeNode *edge = graphs[i]->next;
            while (edge) {
                weights[i][edge->index] = edge->weight;
                edge = edge->next;
            }
        }

        vector<int> gap(graphs.size() + 1, 0);
        gap[0] = graphs.size();
        vector<int> degrees(graphs.size(), 0);
        int result = 0;
        while (degrees[start] != graphs.size()) {
            result += search(graphs, start, end, start, INT_MAX, weights, gap, degrees);
        }
        return result;
    }

private:
    int search(vector<VertexNode *> &graphs, int start, int end, int index, int flow, vector<vector<int>> &weights, vector<int> &gap, vector<int> &degrees) {
        if (index == end) {
            return flow;
        }

        EdgeNode *edge = graphs[index]->next;
        int pivot = flow;
        int border = graphs.size() - 1;
        while (edge) {
            int edgeIndex = edge->index;
            if (weights[index][edgeIndex]) {
                if (degrees[index] == degrees[edgeIndex] + 1) {
                    int value = search(graphs, start, end, edgeIndex, min(pivot, weights[index][edgeIndex]), weights, gap, degrees);
                    weights[index][edgeIndex] -= value;
                    weights[edgeIndex][index] += value;
                    pivot -= value;
                    if (degrees[start] == graphs.size() || !pivot) {
                        return flow - pivot;
                    }
                }
                if (border > degrees[edgeIndex]) {
                    border = degrees[edgeIndex];
                }
            }
            edge = edge->next;
        }

        --gap[degrees[index]];
        if (!~gap[degrees[index]]) {
            degrees[start] = graphs.size();
        } else {
            degrees[index] = border + 1;
            ++gap[degrees[index]];
        }
        return flow - pivot;
    }
};
